aws/sign.go

```diff
@@ -224,37 +239,50 @@ func canonicalQueryString(queryVals url.Values) (string, error) {
 	return strings.Replace(queryVals.Encode(), "+", "%20", -1), nil
 }
 
-func canonicalHeaders(sortedHeaderNames []string, hdr http.Header) (string, error) {
+func canonicalHeaders(sortedHeaderNames []string, host string, hdr http.Header) (string, error) {
 	buffer := new(bytes.Buffer)
 
 	for _, hName := range sortedHeaderNames {
-		canonHdrKey := http.CanonicalHeaderKey(hName)
-		sortedHdrVals := hdr[canonHdrKey]
-		sort.Strings(sortedHdrVals)
-		hdrVals := strings.Join(sortedHdrVals, ",")
+
+		var hdrVals string
+		if hName == "host" {
+			hdrVals = host
+		} else {
+			canonHdrKey := http.CanonicalHeaderKey(hName)
+			sortedHdrVals := hdr[canonHdrKey]
+			sort.Strings(sortedHdrVals)
+			hdrVals = strings.Join(sortedHdrVals, ",")
+		}
+
 		if _, err := fmt.Fprintf(buffer, "%s:%s\n", hName, hdrVals); err != nil {
 			return "", err
 		}
 	}
 
+	// There is intentionally a hanging newline at the end of the
+	// header list.
 	return buffer.String(), nil
 }
 
 // Returns a SHA256 checksum of the request body. Represented as a
 // lowercase hexadecimal string.
 func payloadHash(req *http.Request, hasher hasher) (string, error) {
-	if b, err := ioutil.ReadAll(req.Body); err != nil {
-		return "", err
+	if req.Body != nil {
+		if b, err := ioutil.ReadAll(req.Body); err != nil {
+			return "", err
+		} else {
```

Comment:
I'd pull b, err := ioutil.ReadAll(..) before the if and drop the else clause to reduce the indentation level.